src/main/java/com/amazon/ion/impl/macro/MacroCompiler.kt

```diff
@@ -257,35 +251,48 @@ class MacroCompiler(
     }
 
     /** Moves to the next type and throw [IonException] if it is not the `expected` [IonType]. */
-    private fun IonReaderContinuableCore.nextAndCheckType(expected: IonType, location: String) {
-        confirm(nextValue() != IonCursor.Event.NEEDS_DATA && encodingType == expected) { "$location must be a $expected; found ${encodingType ?: "nothing"}" }
+    private fun ReaderType.nextAndCheckType(expected: IonType, location: String) {
+        confirm(readerNextValue() && readerEncodingType() == expected) { "$location must be a $expected; found ${readerEncodingType() ?: "nothing"}" }
     }
 
     /** Moves to the next type and throw [IonException] if it is not the `expected` [IonType]. */
-    private fun IonReaderContinuableCore.nextAndCheckType(expected0: IonType, expected1: IonType, location: String) {
-        confirm(nextValue() != IonCursor.Event.NEEDS_DATA && (encodingType == expected0 || encodingType == expected1)) { "$location must be a $expected0 or $expected1; found ${encodingType ?: "nothing"}" }
+    private fun ReaderType.nextAndCheckType(expected0: IonType, expected1: IonType, location: String) {
+        confirm(readerNextValue() && (readerEncodingType() == expected0 || readerEncodingType() == expected1)) { "$location must be a $expected0 or $expected1; found ${readerEncodingType() ?: "nothing"}" }
     }
 
     /** Steps into a container, executes [block], and steps out. */
-    private inline fun IonReaderContinuableCore.readContainer(block: IonReaderContinuableCore.() -> Unit) { stepIntoContainer(); block(); stepOutOfContainer() }
+    private inline fun ReaderType.readContainer(block: ReaderType.() -> Unit) { readerStepIntoContainer(); block(); readerStepOutOfContainer() }
 
     /** Executes [block] for each remaining value at the current reader depth. */
-    private inline fun IonReaderContinuableCore.forEachRemaining(block: IonReaderContinuableCore.(IonType) -> Unit) { while (nextValue() != IonCursor.Event.END_CONTAINER) { block(encodingType) } }
+    private inline fun ReaderType.forEachRemaining(block: ReaderType.(IonType) -> Unit) { while (readerNextValue()) { block(readerEncodingType()!!) } }
 
     /** Steps into a container, executes [block] for each value at that reader depth, and steps out. */
-    private inline fun IonReaderContinuableCore.forEachInContainer(block: IonReaderContinuableCore.(IonType) -> Unit) = readContainer { forEachRemaining(block) }
-
-    private fun IonReaderContinuableCore.getTypeAnnotationSymbols(): List<SymbolToken> {
-        if (!hasAnnotations()) {
-            return emptyList()
-        }
-        val annotations = arrayListOf<SymbolToken>()
-        consumeAnnotationTokens { annotations += it }
-        return annotations
-    }
+    private inline fun ReaderType.forEachInContainer(block: ReaderType.(IonType) -> Unit) = readContainer { forEachRemaining(block) }
+
+    abstract fun ReaderType.readerHasAnnotations(): Boolean
```

Comment:
There's no need for these to be extension functions, and I think it may be a little more readable as non-extension functions. (I realize that I made extension functions for `readContainer`, `forEachRemaining`, etc. They are also don't need to be extension functions.) If we get rid of the receiver type on all of these functions, then we no longer need the type parameter on the class. Finally, all of these abstract methods can be `protected`.

Reply:
Done